elastic/kibana
# [Guided onboarding] State management improvements

This PR updates the state management so that we can track the state of multiple guides, and status of each individual step per guide.

Fixes https://github.com/elastic/kibana/issues/138626
Fixes https://github.com/elastic/kibana/issues/140978
Fixes https://github.com/elastic/kibana/issues/139752

### How to test
Setup:
1. Enable guided onboarding via your `kibana.dev.yml` file: `guidedOnboarding.ui: true`
2. Start up Kibana with the example plugins enabled: `yarn start --run-examples`
3. Navigate to `/app/guidedOnboardingExample`
4. Click the "Start search guide" button to trigger the guide ("Search" is currently the only guide that you can walk through its entirety)

Testing scenarios:
- Run through entire "Start search guide" demo flow. Verify "Start using Elastic" button shows after all steps have been completed.
- Refresh the search guide mid-flow and verify state is maintained.
- Start a second guide, i.e., "Observability" and verify panel is updated. Restart the "Search" guide. Verify panel is updated back to the "Search" guide.
- Verify progress bar only shows *after* the first step has been started.
- Verify the step button changes from "Start" to "Continue" if you revisit an in-progress step.

### Summary of changes
The data shape stored in the saved object has been updated as follows:

```
interface GuideState {
  guideId: GuideId; //  one of: search, observability, security
  status: GuideStatus; // one of: in_progress, ready_to_complete, or complete
  isActive?: boolean; // Drives the current guide shown in the dropdown panel
  steps: GuideStep[]; // Array of step IDs with their status
}
```

A saved object will be created per guide. The saved object will only be created if a guide has been started. Only one guide can be "active" at a time. The "active" guide state is what will drive the data in the dropdown panel. The observable only keeps track of the "active" guide.

### What's missing 
There are a couple of issues that are being tracked separately, but somewhat related to state management. Just calling them out here to be explicit about what this PR **does not** include.
- Ability to manually complete a step. I think we can leverage the existing `completeGuideStep` API, although I haven't fully thought this functionality through so there may be other changes required. (https://github.com/elastic/kibana/issues/139780)
- "Quit guide" functionality. I imagine we will want to delete the SO in this case. However, I did not add `delete` support as part of this PR. (https://github.com/elastic/kibana/issues/139804)
